src/Waargonaut/Decode.hs

```diff
@@ -850,14 +893,34 @@ listAt
   => Decoder f a
   -> JCurs
   -> DecodeResult f [a]
-listAt elemD curs = DI.try (down curs) >>= maybe
-  (pure mempty)
-  (rightwardSnoc mempty elemD)
+listAt elemD = withType JT.JsonTypeArray $ \c ->
+  DI.try (down c) >>= maybe (pure mempty) (rightwardSnoc mempty elemD)
 
 -- | Helper function to simplify writing a '[]' decoder.
 list :: Monad f => Decoder f a -> Decoder f [a]
 list d = withCursor (listAt d)
 
+-- | Try to decode an object using the given key and value 'Decoder's at the
+-- given cursor.
+objectAt
```

Comment:
Could we change the name of this function to match the Encoder equivalent of keyValuesAsObject, or something similar. Since an object may have different types on different keys but this only allows a single type?